server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java

```diff
@@ -458,38 +459,62 @@ public void uploadMetadata(
+                segmentInfosSnapshot.write(new ByteBuffersIndexOutput(byteBuffersIndexOutput, "Snapshot of SegmentInfos", "SegmentInfos"));
+                byte[] segmentInfoSnapshotByteArray = byteBuffersIndexOutput.toArrayCopy();
+
+                metadataStreamWrapper.writeStream(
+                    indexOutput,
+                    new RemoteSegmentMetadata(
+                        RemoteSegmentMetadata.fromMapOfStrings(uploadedSegments),
+                        segmentInfoSnapshotByteArray,
+                        segmentInfosSnapshot.getGeneration()
+                    )
+                );
+                indexOutput.close();
+                storeDirectory.sync(Collections.singleton(metadataFilename));
+                remoteMetadataDirectory.copyFrom(storeDirectory, metadataFilename, metadataFilename, IOContext.DEFAULT);
             }
+        } finally {
+            tryAndDeleteLocalFile(metadataFilename, storeDirectory);
+        }
+    }
 
-            ByteBuffersDataOutput byteBuffersIndexOutput = new ByteBuffersDataOutput();
-            segmentInfosSnapshot.write(new ByteBuffersIndexOutput(byteBuffersIndexOutput, "Snapshot of SegmentInfos", "SegmentInfos"));
-            byte[] segmentInfoSnapshotByteArray = byteBuffersIndexOutput.toArrayCopy();
-
-            metadataStreamWrapper.writeStream(
-                indexOutput,
-                new RemoteSegmentMetadata(
-                    RemoteSegmentMetadata.fromMapOfStrings(uploadedSegments),
-                    segmentInfoSnapshotByteArray,
-                    segmentInfosSnapshot.getGeneration()
-                )
-            );
-            indexOutput.close();
-            storeDirectory.sync(Collections.singleton(metadataFilename));
-            remoteMetadataDirectory.copyFrom(storeDirectory, metadataFilename, metadataFilename, IOContext.DEFAULT);
-            storeDirectory.deleteFile(metadataFilename);
+    /**
+     * Generate filename for segment metadata file
+     * @param primaryTerm primary term to be used in the name of metadata file
+     * @param generation commit generation
+     * @return metadata filename
+     */
+    private String getMetadataFileName(long primaryTerm, long generation) {
+        return MetadataFilenameUtils.getMetadataFilename(primaryTerm, generation, this.commonFilenameSuffix);
+    }
+
+    /**
+     * Try to delete file from local store. Fails silently on failures
+     * @param filename: name of the file to be deleted
+     */
+    private void tryAndDeleteLocalFile(String filename, Directory directory) {
+        try {
+            logger.trace("Deleting file: " + filename);
+            directory.deleteFile(filename);
+        } catch (NoSuchFileException | FileNotFoundException e) {
+            logger.trace("Exception while deleting. Missing file : " + filename, e);
+        } catch (IOException e) {
+            logger.warn("Exception while deleting: " + filename, e);
```

Comment:
What would happen in case of `IOException`? Will the next run of `RemoteStoreRefreshListener.afterRefresh()` fail?

Reply:
IOException can happen due to multiple reasons. If the file is still present in the local store, yes the run will fail while creating the file with same name. But that next run which fails would be able to delete the file and the subsequent run can run properly. 

IOException is happening in call runs during delete then all remote refresh will fail until the filename changes

Reply:
Got it.